device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.Client
+{
+    using System;
+    using Microsoft.Azure.Devices.Client.Extensions;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Authentication method that uses a shared access signature token and allows for token refresh. 
+    /// </summary>
+    public abstract class DeviceAuthenticationWithTokenRefresh : IAuthenticationMethod
+    {
+        private string _deviceId;
+        private readonly int _timeToLiveBufferSeconds;
+        private SemaphoreSlim _lock = new SemaphoreSlim(1);
+        private string _token;
+        private DateTime _expiryTime;
+
+        public string DeviceId 
+        {
+            get
+            {
+                return _deviceId;
+            }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceAuthenticationWithTokenRefresh"/> class.
+        /// </summary>
+        /// <param name="deviceId">Device Identifier.</param>
+        public DeviceAuthenticationWithTokenRefresh(string deviceId, 
+            int timeToLiveBufferSeconds = 60) 
```

Comment:
>timeToLiveBufferSeconds [](start = 16, length = 23)

I haven't gotten to where this is used yet. But if this is the SAS Expiry Time, should be 60min to line up with the rest of the SDKs.

Reply:
Two points about this:
1. This is work-in-progress in this PR and isn't hooked-up to actually perform Token refresh. I am currently working on another PR for that. (Safe to ignore in this PR.)
2. Good point about the times. There are two actions here:
- remove the default parameters - bad API practice as they can't be versioned in the future.
- rewrite using the existing timeouts